### PR TITLE
feat(rpc): Asset Unlock status by index

### PR DESCRIPTION
**Pull request by ogabrielides**
## Issue being fixed or feature implemented
Platform in the scope of credit withdrawals, need a way to get the status of an Asset Unlock by index.

## What was done?
A new RPC was created `getassetunlockchainlocks` that accepts Asset Unlock indexes array as parameter and return corresponding status for each index.

The possible outcomes per each index are:
- `chainlocked`: If the Asset Unlock index is mined on a Chainlocked block.
- `mined`: If no Chainlock information is available, and the Asset Unlock index is mined.
- `mempooled`: If the Asset Unlock index is in the mempool.
- `unknown`: If none of the above are valid.

Note: This RPC is whitelisted for the Platform RPC user.

## How Has This Been Tested?
Inserted on `feature_asset_locks.py` covering cases where Asset Unlock txs are in mempool, mined and not present.

## Breaking Changes
no

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -336,6 +339,120 @@ static UniValue gettxchainlocks(const JSONRPCRequest& request)
...
+                             {
+                                     {RPCResult::Type::STR, "", "Status of the Asset Unlock index: {chainlocked|mempooled|null}"},
+                             }},
+                    }
+            },
+            RPCExamples{
+                    HelpExampleCli("getassetunlockchainlocks", "'[\"myindex\",...]'")
+                    + HelpExampleRpc("getassetunlockchainlocks", "[\"myindex\",...]")
+            },
+    }.Check(request);
```
Comment (PastaPastaPasta) [inline src/rpc/rawtransaction.cpp:367]: Have this in a separate method

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -336,6 +339,120 @@ static UniValue gettxchainlocks(const JSONRPCRequest& request)
...
+
+    if (g_txindex) {
+        g_txindex->BlockUntilSyncedToCurrentChain();
+    }
+
+    CBlockIndex* pTipBlockIndex{nullptr};
+    {
+        LOCK(cs_main);
+        pTipBlockIndex = chainman.ActiveChain().Tip();
+    }
```
Comment (PastaPastaPasta) [inline src/rpc/rawtransaction.cpp:388]: refactor to use WITH_LOCK
```suggestion
    CBlockIndex* pTipBlockIndex= WITH_LOCK(cs_main, return chainman.ActiveChain().Tip());
```

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -336,6 +339,120 @@ static UniValue gettxchainlocks(const JSONRPCRequest& request)
...
+
+    CBlockIndex* pBlockIndex{nullptr};
+    bool chainlock_info = false;
+    llmq::CChainLockSig bestclsig = llmq_ctx.clhandler->GetBestChainLock();
+    if (bestclsig.IsNull()) {
+        // If no CL info is available, try to use CbTx CL information
+        auto cbtx_best_cl = GetNonNullCoinbaseChainlock(pTipBlockIndex);
+        if (cbtx_best_cl.has_value()) {
```
Comment (PastaPastaPasta) [inline src/rpc/rawtransaction.cpp:396]: if-init

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -336,6 +339,120 @@ static UniValue gettxchainlocks(const JSONRPCRequest& request)
...
+    CBlockIndex* pTipBlockIndex{nullptr};
+    {
+        LOCK(cs_main);
+        pTipBlockIndex = chainman.ActiveChain().Tip();
+    }
+
+    CBlockIndex* pBlockIndex{nullptr};
+    bool chainlock_info = false;
+    llmq::CChainLockSig bestclsig = llmq_ctx.clhandler->GetBestChainLock();
+    if (bestclsig.IsNull()) {
```
Comment (PastaPastaPasta) [inline src/rpc/rawtransaction.cpp:393]: we should create a separate feat PR which on ProcessBlock we will submit a blocks chain lock to the chain lock system

File: src/rpc/rawtransaction.cpp [hunk 5]
```diff
@@ -336,6 +339,120 @@ static UniValue gettxchainlocks(const JSONRPCRequest& request)
...
+                }
+            }
+            if (is_mempooled)
+                obj.pushKV(str_indexes[i].get_str(), "mempooled");
+            else {
+                UniValue jnull(UniValue::VNULL);
+                obj.pushKV(str_indexes[i].get_str(), jnull);
+            }
+            result_arr.push_back(obj);
+        }
```
Comment (PastaPastaPasta) [inline src/rpc/rawtransaction.cpp:450]: redundant scope

**Review by thephez — COMMENTED** [review]
What about calling it `assetunlockstatus` instead since it returns one of several status values?

File: src/rpc/rawtransaction.cpp (hunk 4 above)
Comment (ogabrielides) [reply src/rpc/rawtransaction.cpp:393]: https://github.com/dashpay/dash/pull/5779

File: src/rpc/rawtransaction.cpp (hunk 1 above)
Comment (ogabrielides) [reply src/rpc/rawtransaction.cpp]: OK (Just no other RPC has a separate method on this file fyi)

**Review by UdjinM6 — DISMISSED** [review]
utACK

File: doc/release-notes-5776.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
+Added RPC
+--------
+
+- `getassetunlockchainlocks` RPC allows to fetch Asset Unlock txs by their withdrawal index. The RPC accepts an array of indexes and returns status for each index.
+The possible outcomes per each index are:
+- "chainlocked": If the Asset Unlock index is mined on a Chainlocked block.
+- "mined": If no Chainlock information is available, and the Asset Unlock index is mined.
+- "mempooled": If the Asset Unlock index is in the mempool.
+- null: If none of the above are valid.
```
Comment (thephez) [inline doc/release-notes-5776.md:9]:
```suggestion
- `getassetunlockstatuses` RPC allows fetching of Asset Unlock txs by their withdrawal index. The RPC accepts an array of indexes and returns status for each index.
The possible outcomes per each index are:
- "chainlocked": If the Asset Unlock index is mined on a ChainLocked block.
- "mined": If no ChainLock information is available, and the Asset Unlock index is mined.
- "mempooled": If the Asset Unlock index is in the mempool.
- null: If none of the above are valid.
```

**Review by thephez — COMMENTED** [review]
Probably should update the release notes now too so they match

**Review by UdjinM6 — DISMISSED** [review]
re-utACK

File: test/functional/feature_asset_locks.py [hunk 7]
```diff
@@ -330,6 +330,16 @@ def run_test(self):
 
         txid = self.send_tx(asset_unlock_tx)
         assert "assetUnlockTx" in node.getrawtransaction(txid, 1)
+
+        indexes_statuses = self.nodes[0].getassetunlockstatuses(["101", "102", "300"])
+        self.log.info(f'{indexes_statuses}')
+        assert any('101' in d for d in indexes_statuses)
```
Comment (knst) [inline test/functional/feature_asset_locks.py:336]: why not simple `assert_equal(indexes_statuses, ["mempooled", None, None])` ?

File: src/rpc/rawtransaction.cpp [hunk 8]
```diff
@@ -336,6 +339,119 @@ static UniValue gettxchainlocks(const JSONRPCRequest& request)
...
+    CCreditPool pool = node.creditPoolManager->GetCreditPool(pBlockIndex, Params().GetConsensus());
+
+    for (const auto i : irange::range(str_indexes.size())) {
+        UniValue obj(UniValue::VOBJ);
+        uint64_t index{};
+        if (!ParseUInt64(str_indexes[i].get_str(), &index)) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "invalid index");
+        }
+        if (pool.indexes.Contains(index)) {
+            obj.pushKV(str_indexes[i].get_str(), chainlock_info ? "chainlocked" : "mined");
```
Comment (knst) [inline src/rpc/rawtransaction.cpp:424]: consider using 2 credit-pools such as: https://github.com/dashpay/dash/commit/4a18b8450b61aa812ed1c14b3a567b59fe4a9a65

File: test/functional/feature_asset_locks.py (hunk 7 above)
Comment (ogabrielides) [reply test/functional/feature_asset_locks.py:336]: Well, I want to make sure that '101' is also returned.

**Review by thephez — APPROVED** [review]

File: test/functional/feature_asset_locks.py [hunk 9]
```diff
@@ -330,6 +330,16 @@ def run_test(self):
...
+        assert_equal(next((d['101'] for d in indexes_statuses if '101' in d), None), "mempooled")
+        assert any('102' in d for d in indexes_statuses)
+        assert_equal(next((d['102'] for d in indexes_statuses if '102' in d), None), None)
+        assert any('300' in d for d in indexes_statuses)
+        assert_equal(next((d['300'] for d in indexes_statuses if '300' in d), None), None)
```
Comment (knst) [inline test/functional/feature_asset_locks.py:341]:
```suggestion
         assert_equal([{'101': 'mempooled'}, {'102': None}, {'300': None}], indexes_statuses)
```

**Review by knst — DISMISSED** [review]
ACK

File: src/rpc/rawtransaction.cpp [hunk 10]
```diff
@@ -441,10 +443,10 @@ static UniValue getassetunlockstatuses(const JSONRPCRequest& request)
             }
         }
         if (is_mempooled)
-            obj.pushKV(str_indexes[i].get_str(), "mempooled");
+            obj.pushKV("status", "mempooled");
         else {
             UniValue jnull(UniValue::VNULL);
-            obj.pushKV(str_indexes[i].get_str(), jnull);
+            obj.pushKV("status", jnull);
```
Comment (knst) [inline src/rpc/rawtransaction.cpp:449]: nit: `obj.pushKV("status", UniValue{});`  does same, doesn't it?

**Review by knst — DISMISSED** [review]
utACK

File: src/init.cpp [hunk 11]
```diff
@@ -1924,6 +1924,8 @@ bool AppInitMain(const CoreContext& context, NodeContext& node, interfaces::Bloc
                 node.evodb = std::make_unique<CEvoDB>(nEvoDbCache, false, fReset || fReindexChainState);
                 node.mnhf_manager.reset();
                 node.mnhf_manager = std::make_unique<CMNHFManager>(*node.evodb);
+                node.creditPoolManager.reset();
+                node.creditPoolManager = std::make_unique<CCreditPoolManager>(*node.evodb);
```
Comment (PastaPastaPasta) [inline src/init.cpp:1928]: Why is this here? Is it an artifact of a rebase or something?

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
Please see the following patch; note this changes behavior to not return a null but "unknown"

```
Subject: [PATCH] lots of lamdas
---
Index: src/rpc/rawtransaction.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/rpc/rawtransaction.cpp b/src/rpc/rawtransaction.cpp
--- a/src/rpc/rawtransaction.cpp	(revision 22dd732df77f69163293b8adcea1e9af1f3c6c73)
+++ b/src/rpc/rawtransaction.cpp	(date 1703220226644)
@@ -388,25 +388,31 @@
     }
 
     CBlockIndex* pTipBlockIndex{WITH_LOCK(cs_main, return chainman.ActiveChain().Tip())};
-
-    CBlockIndex* pBlockIndexBestCL{nullptr};
-    const llmq::CChainLockSig bestclsig{llmq_ctx.clhandler->GetBestChainLock()};
-    if (bestclsig.IsNull()) {
-        // If no CL info is available, try to use CbTx CL information
-        if (const auto cbtx_best_cl = GetNonNullCoinbaseChainlock(pTipBlockIndex); cbtx_best_cl.has_value()) {
-            pBlockIndexBestCL = pTipBlockIndex->GetAncestor(pTipBlockIndex->nHeight - cbtx_best_cl->second - 1);
-        }
-    }
-
     if (!pTipBlockIndex) {
         throw JSONRPCError(RPC_INTERNAL_ERROR, "No blocks in chain");
     }
 
+    const auto pBlockIndexBestCL = [&]() -> const CBlockIndex* {
+        if (llmq_ctx.clhandler->GetBestChainLock().IsNull()) {
+            // If no CL info is available, try to use CbTx CL information
+            if (const auto cbtx_best_cl = GetNonNullCoinbaseChainlock(pTipBlockIndex)) {
+                return pTipBlockIndex->GetAncestor(pTipBlockIndex->nHeight - cbtx_best_cl->second - 1);
+            }
+        }
+        return nullptr;
+    }();
+
     // We need in 2 credit pools: at tip of chain and on best CL to know if tx is mined or chainlocked
     // Sometimes that's two different blocks, sometimes not and we need to initialize 2nd creditPoolManager
-    std::optional<CCreditPool> poolCL;
-    if (pBlockIndexBestCL != nullptr) poolCL = node.creditPoolManager->GetCreditPool(pBlockIndexBestCL, Params().GetConsensus());
-    std::optional<CCreditPool> poolOnTip{};
+    std::optional<CCreditPool> poolCL = pBlockIndexBestCL ?
+            std::make_optional(node.creditPoolManager->GetCreditPool(pBlockIndexBestCL, Params().GetConsensus())) :
+            std::nullopt;
+    auto poolOnTip = [&]() -> std::optional<CCreditPool> {
+        if (pTipBlockIndex != pBlockIndexBestCL) {
+            return std::make_optional(node.creditPoolManager->GetCreditPool(pTipBlockIndex, Params().GetConsensus()));
+        }
+        return std::nullopt;
+    }();
 
     for (const auto i : irange::range(str_indexes.size())) {
         UniValue obj(UniValue::VOBJ);
@@ -415,39 +421,32 @@
             throw JSONRPCError(RPC_INVALID_PARAMETER, "invalid index");
         }
         obj.pushKV("index", index);
-        if (poolCL.has_value() && poolCL->indexes.Contains(index)) {
-            obj.pushKV("status", "chainlocked");
-            result_arr.push_back(obj);
-            continue;
-        }
-        if (pTipBlockIndex != pBlockIndexBestCL) {
-            if (!poolOnTip.has_value()) poolOnTip = node.creditPoolManager->GetCreditPool(pTipBlockIndex, Params().GetConsensus());
-            if (poolOnTip->indexes.Contains(index)) {
-                obj.pushKV("status", "mined");
-                result_arr.push_back(obj);
-                continue;
+
+        auto status_to_push = [&]() -> std::string {
+            if (poolCL.has_value() && poolCL->indexes.Contains(index)) {
+                return "chainlocked";
+            }
+            if (poolOnTip.has_value() && poolOnTip->indexes.Contains(index)) {
+                return "mined";
             }
-        }
-        LOCK(mempool.cs);
-        bool is_mempooled = false;
-        for (const CTxMemPoolEntry& e : mempool.mapTx) {
-            if (e.GetTx().nType == CAssetUnlockPayload::SPECIALTX_TYPE) {
-                CAssetUnlockPayload assetUnlockTx;
-                if (!GetTxPayload(e.GetTx(), assetUnlockTx)) {
-                    throw JSONRPCError(RPC_TRANSACTION_ERROR, "bad-assetunlocktx-payload");
-                }
-                if (index == assetUnlockTx.getIndex()) {
-                    is_mempooled = true;
-                    break;
-                }
-            }
-        }
-        if (is_mempooled)
-            obj.pushKV("status", "mempooled");
-        else {
-            UniValue jnull(UniValue::VNULL);
-            obj.pushKV("status", jnull);
-        }
+            bool is_mempooled = [&]() {
+                LOCK(mempool.cs);
+                return std::any_of(mempool.mapTx.begin(), mempool.mapTx.end(), [index](const CTxMemPoolEntry &e) {
+                    if (e.GetTx().nType == CAssetUnlockPayload::SPECIALTX_TYPE) {
+                        if (CAssetUnlockPayload assetUnlockTx; GetTxPayload(e.GetTx(), assetUnlockTx)) {
+                            return index == assetUnlockTx.getIndex();
+                        } else {
+                            throw JSONRPCError(RPC_TRANSACTION_ERROR, "bad-assetunlocktx-payload");
+                        }
+                    }
+                    return false;
+                });
+            }();
+            return is_mempooled ? "mempooled" : "unknown";
+        };
+
+        obj.pushKV("status", status_to_push());
+
         result_arr.push_back(obj);
     }
 
```

File: src/init.cpp (hunk 11 above)
Comment (knst) [reply src/init.cpp:1928]: that's indeed duplicate of re-init, see https://github.com/dashpay/dash/pull/5776/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dL1943

```suggestion
```

Comment (ogabrielides) [reply src/init.cpp:1928]: No, this doesn't work. RPC needs to access to `node.creditPoolManager`

Comment (knst) [reply src/init.cpp:1928]: indeed. Good catch! Though, 2 objects should not co-exist in this cruel world.
here's hot fix: https://github.com/knst/dash/commit/648ac85cea4b5f15ebc4be8ff20b37b90fc02ee0
Proper fix requires deglobalization of `creditPoolManager`, out-of-score for this PR

File: src/init.cpp [hunk 12]
```diff
@@ -1924,8 +1925,7 @@ bool AppInitMain(const CoreContext& context, NodeContext& node, interfaces::Bloc
                 node.evodb = std::make_unique<CEvoDB>(nEvoDbCache, false, fReset || fReindexChainState);
                 node.mnhf_manager.reset();
                 node.mnhf_manager = std::make_unique<CMNHFManager>(*node.evodb);
-                node.creditPoolManager.reset();
-                node.creditPoolManager = std::make_unique<CCreditPoolManager>(*node.evodb);
+
```
Comment (knst) [inline src/init.cpp:1928]: nit: extra \n

**Review by knst — APPROVED** [review]
utACK

**Review by UdjinM6 — DISMISSED** [review]
👍 

utACK

File: doc/release-notes-5776.md [hunk 13]
```diff
@@ -0,0 +1,11 @@
+Added RPC
+--------
+
+- `getassetunlockstatuses` RPC allows fetching of Asset Unlock txs by their withdrawal index. The RPC accepts an array of indexes and returns status for each index.
+The possible outcomes per each index are:
+- "chainlocked": If the Asset Unlock index is mined on a ChainLocked block.
+- "mined": If no ChainLock information is available, and the Asset Unlock index is mined.
+- "mempooled": If the Asset Unlock index is in the mempool.
+- "unknown"": If none of the above are valid.
```
Comment (PastaPastaPasta) [inline doc/release-notes-5776.md:9]: drop extra "
```suggestion
- "unknown": If none of the above are valid.
```

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

**Review by UdjinM6 — APPROVED** [review]
utACK
